packages/components/src/internal/components/editable/EditableGrid.tsx

```diff
@@ -701,36 +711,64 @@ export class EditableGrid extends ReactN.PureComponent<EditableGridProps, Editab
     }
 
     bulkAdd = (data: OrderedMap<string, any>): Promise<any> => {
-        const { addControlProps, bulkAddProps } = this.props;
+        const { addControlProps, bulkAddProps, onBulkAdd  } = this.props;
         const { nounSingular, nounPlural } = addControlProps;
         const model = this.getModel(this.props);
 
         const numItems = data.get('numItems');
-        let updatedData = data.delete('numItems');
 
-        if (numItems) {
-            if (bulkAddProps.columnFilter) {
-                updatedData = this.restoreBulkInsertData(model, updatedData);
-            }
-
-            return new Promise(resolve => {
-                addRows(model, numItems, updatedData);
-                this.onRowCountChange();
-                resolve({
-                    success: true,
-                    message: 'Added ' + numItems + ' ' + (numItems > 1 ? nounPlural : nounSingular),
+        if (!numItems) {
+            return new Promise((resolve, reject) => {
+                reject({
+                    exception: 'Quantity unknown.  No ' + nounPlural + ' added.',
                 });
             });
         }
-        return new Promise((resolve, reject) => {
-            reject({
-                exception: 'Quantity unknown.  No ' + nounPlural + ' added.',
+
+        let pivotKey = undefined;
+        let pivotValues = []
+        let totalItems = 0;
+        if (onBulkAdd) {
+            const bulkAddData  = onBulkAdd(data);
+            if (bulkAddData.validationMsg) {
+                return new Promise((resolve, reject) => {
+                    reject({
+                        exception: bulkAddData.validationMsg,
+                    });
+                });
+            }
+            totalItems = bulkAddData.totalItems;
+            pivotKey = bulkAddData.pivotKey;
+            pivotValues = bulkAddData.pivotValues;
```

Comment:
looks like these are just used in one place below (and used together), could we then instead just hold onto the `bulkAddData` object and use it for each of these values below at line 755?